src/node_osrm.cpp

```diff
@@ -1,1161 +1,173 @@
-#include <exception>
-#include <string>
+
+#include "node_osrm.hpp"
+#include "node_osrm_support.hpp"
 
 namespace node_osrm
 {
 
-using osrm_ptr = std::unique_ptr<osrm::OSRM>;
-using engine_config_ptr = std::unique_ptr<osrm::EngineConfig>;
-using route_parameters_ptr = std::unique_ptr<osrm::RouteParameters>;
-using trip_parameters_ptr = std::unique_ptr<osrm::TripParameters>;
-using tile_parameters_ptr = std::unique_ptr<osrm::TileParameters>;
-using match_parameters_ptr = std::unique_ptr<osrm::MatchParameters>;
-using nearest_parameters_ptr = std::unique_ptr<osrm::NearestParameters>;
-using table_parameters_ptr = std::unique_ptr<osrm::TableParameters>;
-namespace
-{
-template <class T, class... Types> std::unique_ptr<T> make_unique(Types &&... Args)
-{
-    return (std::unique_ptr<T>(new T(std::forward<Types>(Args)...)));
-}
-}
+Engine::Engine(osrm::EngineConfig &config) : Base(), this_(std::make_shared<osrm::OSRM>(config)) {}
 
-// Supports
-engine_config_ptr argumentsToEngineConfig(const Nan::FunctionCallbackInfo<v8::Value> &args)
+Nan::Persistent<v8::Function> &Engine::constructor()
 {
-    Nan::HandleScope scope;
-    auto engine_config = make_unique<osrm::EngineConfig>();
-
-    if (args.Length() == 0)
-    {
-        return engine_config;
-    }
-    else if (args.Length() > 1)
-    {
-        Nan::ThrowError("Only accepts one parameter");
-        return engine_config_ptr();
-    }
-
-    BOOST_ASSERT(args.Length() == 1);
-
-    if (args[0]->IsString())
-    {
-        engine_config->storage_config = osrm::StorageConfig(
-            *v8::String::Utf8Value(Nan::To<v8::String>(args[0]).ToLocalChecked()));
-        engine_config->use_shared_memory = false;
-        return engine_config;
-    }
-    else if (!args[0]->IsObject())
-    {
-        Nan::ThrowError("Parameter must be a path or options object");
-        return engine_config_ptr();
-    }
-
-    BOOST_ASSERT(args[0]->IsObject());
-    auto params = Nan::To<v8::Object>(args[0]).ToLocalChecked();
-
-    auto path = params->Get(Nan::New("path").ToLocalChecked());
-    auto shared_memory = params->Get(Nan::New("shared_memory").ToLocalChecked());
-    if (!path->IsUndefined())
-    {
-        engine_config->storage_config = osrm::StorageConfig(
-            *v8::String::Utf8Value(Nan::To<v8::String>(path).ToLocalChecked()));
-    }
-    if (!shared_memory->IsUndefined())
-    {
-        if (shared_memory->IsBoolean())
-        {
-            engine_config->use_shared_memory = Nan::To<bool>(shared_memory).FromJust();
-        }
-        else
-        {
-            Nan::ThrowError("Shared_memory option must be a boolean");
-            return engine_config_ptr();
-        }
-    }
-
-    if (path->IsUndefined() && !engine_config->use_shared_memory)
-    {
-        Nan::ThrowError("Shared_memory must be enabled if no path is "
-                        "specified");
-        return engine_config_ptr();
-    }
-
-    return engine_config;
+    static Nan::Persistent<v8::Function> init;
```

Comment:
This function is needed for the thread-safe singleton impl. without relying on global constructors.